### PR TITLE
AP_HAL - RingBuffer: Fixed off by one in space() method.

### PR DESCRIPTION
**Pull request by danorbtnk**
Whenever the head and tail indexes were equal space() would be off by
one. This happens in particular right after construction.

### CODE REVIEW

**Comment by lucasdemarchi** [comment]
Can you update the unit test accordingly?

**Comment by danorbtnk** [comment]
I'll look into it.

**Comment by danorbtnk** [comment]
I can't even find it?

**Comment by lucasdemarchi** [comment]
oh... it was never pushed to upstream, dunno why. And it was not a unit test but rather a stress test to get race conditions. I was talking about this one here: https://github.com/mbelluzzo/ardupilot/commit/eab3f79ff5872bc1f6f115cee02885ed9687952a

@mbelluzzo any idea why this hasn't been sent?

**Comment by mbelluzzo** [comment]
I think it wasn't sent because it wasn't an unit test. But I think it is very useful. I will send it so ppl can properly review it.

**Comment by mbelluzzo** [comment]
https://github.com/ArduPilot/ardupilot/pull/6714

File: libraries/AP_HAL/utility/RingBuffer.cpp [hunk 1]
```diff
@@ -66,7 +66,9 @@ uint32_t ByteBuffer::space(void) const
         ret = size;
     }
 
-    ret += _head - tail - 1;
+	if(_head != tail) {
```
Comment (mbelluzzo) [inline libraries/AP_HAL/utility/RingBuffer.cpp]: This should always be true because if _head equals to tail, than the previous if would have returned already, wouldn't?

**Comment by magicrub** [comment]
I'm closing this as "not broken". As @mbelluzzo said, this new condition is always true due to the above <= check so the existing code should work correctly. @danorbtnk did you see evidence of this not working? If so, please reopen to discuss.

**Comment by lucasdemarchi** [comment]
With the exception that in the line above it sets ret instead of returning.  Let's leave this open for a while and see if we can reproduce the issue .

**Comment by magicrub** [comment]
ahh yes, we should probably return after the res = size;

**Comment by mbelluzzo** [comment]
Yep, I was biased by the original code that had the entire function in 1 statement (that works kkk).

**Comment by OXINARF** [comment]
I'm a bit surprised by the quick finger in a PR touching such a critical piece of code. PRs like this need to be properly analyzed so we neither merge a bug nor we miss a bugfix. This PR comes after a discussion I had today in Gitter with @danorbtnk where he was testing the implementation and getting problems.

Interestingly I've lost time looking at this piece of code before and deemed it was correct. Now, after giving a good thought about this, I was already prepared to do a new PR with a commit I thought was the correct fix. Fortunately, before I did, it hit me why the current code is actually correct. My fix was going to simply remove the `- 1` from the expression (which this PR also does, but only in the case where the buffer is empty), but that subtraction is actually correct, for a simple reason: `head` always points to the first byte where to read and `tail` always points to first byte where to write. This means that the buffer always needs an empty position where `tail` can point to. There is then the special case where the buffer is empty and `head` and `tail` point to the same place. If we allowed the buffer to use the full size, `head` and `tail` would point to the same place when empty and when full and we wouldn't know which was true.

I'm sure the implementation can be fixed by either internally have the asked `size` + 1 allocated or changing where `tail` points to, but I don't think it is worth losing much time with it. Adding comments explaining it might be though.

@danorbtnk When I said today that it looked like it wasn't losing data, I was wrong in a way. Although the buffer didn't lose data, it wasn't allowing a write to its designed size - because you didn't check the `write` return value, it looked like it wrote everything when it didn't. This is something your sample code mentioned in #6701 also does, it doesn't check if the entire string was written to the UART.

**Comment by danorbtnk** [comment]
@OXINARF Your scenario doesn't happen, let's analyze the edge cases:

* case 1: `head == 0 && tail == size - 1`. The first if statement is triggered since `head <= tail`, so ret is initally set to `size`. Then the substraction occurs, `head - tail - 1 == -(size - 1) - 1 == - size`. So the buffer is full once tail is at index size - 1 and tail won't be moved to head.
* case 2: `head > tail`, i.e. tail has wrapped around. The if statement will not be triggered and `ret == 0` initially. Now the edge case `head == tail + 1`, it follows that `head - tail  -1 == 1 - 1 == 0`, so again tail will not be moved onto head.

In conclusion in the current implementation `head == tail` should _only_ happen if the buffer is empty. Otherwise there is inconsistent behavior, i.e. more bugs.

**Comment by lucasdemarchi** [comment]
Let's leave this open for a while and see if we can reproduce the issue. Please.

**Comment by lucasdemarchi** [comment]
Btw, this could be checked on a unit test.  @danorbtnk if I add the infra to handle RingBuffer, can you add the test to check your case?

**Comment by danorbtnk** [comment]
@lucasdemarchi sure. I already had code reproducing the issue though: https://gist.github.com/danorbtnk/95fb6c68806cd0fff328c22e149b37da I'll gladly also upload the final code that tracked the issue down to the space() method.

**Comment by danorbtnk** [comment]
Note that current bytebuffer code uses `head == tail` as its empty check: https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL/utility/RingBuffer.cpp#L74

**Comment by OXINARF** [comment]
Like I said to @danorbtnk in Gitter, I wrote my post at 4AM, but I've read it again and it looks clear to me. It certainly wasn't to @danorbtnk though.

What I said was that, by design, **`head` can't equal `tail`** (except in the special case of an empty buffer). This means that we always need to have one empty space. And if we always need one empty space, the available space to write (i.e. `space()`) is always going to be `size - 1 - number_of_written_bytes`. Let me give a simple example to be completely clear; a buffer set with size 5:

- initially head and tail are both 0, `available` returns 0 and `space` returns 4
- then you write 3 bytes, head is 0, tail is 3, `available` returns 3 and `space` returns 1
- then you write 1 byte, head is 0, tail is 4, `available` returns 4 and `space` returns 0; you can no longer write more bytes until it is read
- you read 3 bytes, head is 3, tail is 4, `available` returns 1 and `space` returns 3
- you write 3 bytes, head is 3, tail is 2, `available` returns 4 and `space` returns 0
<br/>

- clear the buffer, head and tail are both 0, `available` returns 0 and `space` returns 4
- **try** to write 5 bytes, head is 0, tail is 4, `available` returns 4 and `space` returns 0; `write` will return 4, meaning one byte couldn't be written

@lucasdemarchi Reproduce what? The only issue here is that it isn't documented that from the requested `size`, you can only fill the buffer with up to `size - 1` bytes (due to the implementation). So, in a single thread, at any given time, we'll have `available() + space() == size - 1` instead of the more expected `available() + space() == size`.

I'm closing this (again) for three reasons:

- this doesn't fix the API to have a `available() + space() == size` semantic (@danorbtnk said he was looking to do it and a new PR with it is welcome)
- this makes it worse (and unacceptable), by making the API behave differently when the buffer is empty or not
- I'm sure of what I said above (which is why I closed it the first time)
